### ISSUE TITLE
Segmentation fault when writing empty VLEN data

### ISSUE THREAD

**Issue by colingavin** [h5py/h5py#1245]
The following segfaults for me on macOS (10.12.6) with python 3.7.3 from pyenv.
```
import h5py
import numpy as np

print(h5py.version.info)
d = np.core.records.fromarrays([[], []], names='a,b', formats='|V16,O')
print('array: ', repr(d))
f = h5py.File('tmp.h5', 'w')
f.create_dataset('test', data=d, dtype=[('a', '|V16'), ('b', h5py.special_dtype(vlen=np.float_))])
```
The output is:
```
Summary of the h5py configuration
---------------------------------

h5py    2.9.0
HDF5    1.10.4
Python  3.7.3 (default, Mar 27 2019, 09:23:32) 
[Clang 9.0.0 (clang-900.0.39.2)]
sys.platform    darwin
sys.maxsize     9223372036854775807
numpy   1.16.4

array:  rec.array([],
          dtype=[('a', 'V16'), ('b', 'O')])
Segmentation fault: 11
```
Note that replacing `d` with the following does not crash and writes valid data:
```
b = np.tile(None, 1)
b[0] = np.array([1.0])
d = np.core.records.fromarrays([[np.array(b'\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0', dtype='|V16')], b], names='a,b', formats='|V16,O')
```
There does not appear to be anything special about the length of the byte string, but the crash goes away if it is removed entirely (so the dtype is `['b', 'O']`).

**Comment by scopatz** [comment]
There probably isn't a check for zero-sized data and so it is trying to deref the array location.  PRs welcome!

**Comment by colingavin** [comment]
A little more research: the crash is at https://github.com/h5py/h5py/blob/7d50c06ea42c9238a2659f501fae4816ad4c4fd2/h5py/_conv.pyx#L745
in `ndarray2vlen` when the number of elements `nl` is zero. It appears that in this case the `void* buf_i` being passed in is not a valid `PyObject**`.

This call is originating from https://github.com/h5py/h5py/blob/7d50c06ea42c9238a2659f501fae4816ad4c4fd2/h5py/_proxy.pyx#L163
I don't know the internals of h5py well enough to know whether the problem is the allocation of `conv_buf` at the original call site, or whether the zero-length case should be handled in the conversion function.

**Comment by scopatz** [comment]
I think handling it in the conv.pyx might be the right way to go here.

**Comment by takluyver** [comment]
Closed by #1253.
